File: backend/src/database/repositories/payment.py
"""
Payment Repository

Repository for Payment model with CRUD operations.
"""

from typing import Any
from sqlalchemy import select
from database.types import Repository, DatabaseAdapter
from database.models.payment import Payment


class PaymentRepository(Repository[Payment]):
    """
    Payment Repository

    Provides CRUD operations for Payment model.
    Payments don't have soft delete (audit trail must be permanent).
    """

    def __init__(self, adapter: DatabaseAdapter):
        super().__init__(adapter)

    async def find_by_id(self, id: str, include_deleted: bool = False) -> Payment | None:
        """
        Find payment by ID

        Args:
            id: Payment UUID
            include_deleted: Ignored for payments (no soft delete)

        Returns:
            Payment instance or None
        """
        async with await self.adapter.get_session() as session:
            result = await session.execute(select(Payment).where(Payment.payment_id == id))
            return result.scalar_one_or_none()

    async def find_by_stripe_session_id(self, session_id: str) -> Payment | None:
        """
        Find payment by Stripe Checkout Session ID

        Args:
            session_id: Stripe session ID

        Returns:
            Payment instance or None
        """
        async with await self.adapter.get_session() as session:
            result = await session.execute(
                select(Payment).where(Payment.stripe_checkout_session_id == session_id)
            )
            return result.scalar_one_or_none()

    async def find_by_stripe_payment_intent_id(self, payment_intent_id: str) -> Payment | None:
        """
        Find payment by Stripe Payment Intent ID

        Args:
            payment_intent_id: Stripe payment intent ID

        Returns:
            Payment instance or None
        """
        async with await self.adapter.get_session() as session:
            result = await session.execute(
                select(Payment).where(Payment.stripe_payment_intent_id == payment_intent_id)
            )
            return result.scalar_one_or_none()

    async def find_by_user(
        self, user_id: str, skip: int = 0, limit: int = 100, include_deleted: bool = False
    ) -> list[Payment]:
        """
        Find all payments for a user

        Args:
            user_id: User UUID
            skip: Number of records to skip
            limit: Maximum number of records to return
            include_deleted: Ignored for payments

        Returns:
            List of Payment instances
        """
        async with await self.adapter.get_session() as session:
            query = (
                select(Payment)
                .where(Payment.user_id == user_id)
                .offset(skip)
                .limit(limit)
                .order_by(Payment.created_at.desc())
            )

            result = await session.execute(query)
            return list(result.scalars().all())

    async def find_all(
        self, skip: int = 0, limit: int = 100, include_deleted: bool = False
    ) -> list[Payment]:
        """
        Find all payments with pagination

        Args:
            skip: Number of records to skip
            limit: Maximum number of records to return
            include_deleted: Ignored for payments

        Returns:
            List of Payment instances
        """
        async with await self.adapter.get_session() as session:
            query = select(Payment).offset(skip).limit(limit).order_by(Payment.created_at.desc())

            result = await session.execute(query)
            return list(result.scalars().all())

    async def create(self, data: dict[str, Any]) -> Payment:
        """
        Create new payment

        Args:
            data: Payment data dictionary

        Returns:
            Created Payment instance
        """
        async with await self.adapter.get_session() as session:
            payment = Payment(**data)
            session.add(payment)
            await session.commit()
            await session.refresh(payment)
            return payment

    async def update(self, id: str, data: dict[str, Any]) -> Payment | None:
        """
        Update payment

        Args:
            id: Payment UUID
            data: Updated data dictionary

        Returns:
            Updated Payment instance or None
        """
        async with await self.adapter.get_session() as session:
            payment = await self.find_by_id(id)
            if not payment:
                return None

            for key, value in data.items():
                if hasattr(payment, key):
                    setattr(payment, key, value)

            session.add(payment)
            await session.commit()
            await session.refresh(payment)
            return payment

    async def soft_delete(self, id: str) -> Payment | None:
        """
        Soft delete not applicable for payments (audit trail)

        Args:
            id: Payment UUID

        Returns:
            None (not implemented)
        """
        # Payments don't support soft delete (audit requirement)
        return None

    async def hard_delete(self, id: str) -> bool:
        """
        Hard delete not recommended for payments (audit trail)

        Args:
            id: Payment UUID

        Returns:
            False (not recommended to delete payments)
        """
        # Payments should not be deleted (compliance/audit requirement)
        return False

    async def restore(self, id: str) -> Payment | None:
        """
        Restore not applicable for payments

        Args:
            id: Payment UUID

        Returns:
            None (not implemented)
        """
        # Payments don't support soft delete
        return None

File: shared/database/src/repositories/payment.ts
/**
 * Payment repository
 * Handles payment CRUD operations
 */

import { BaseRepository } from './base';
import { Payment, mapToPayment } from '../types';

/**
 * Payment repository implementation
 * Note: Payments do not have soft delete
 */
export class PaymentRepository extends BaseRepository {
  /**
   * Find payment by ID
   */
  async findById(paymentId: string, userId: string): Promise<Payment | null> {
    const { rows } = await this.db.query<any>(
      `SELECT * FROM payments
       WHERE payment_id = $1
         AND user_id = $2`,
      [paymentId, userId],
    );

    return rows[0] ? mapToPayment(rows[0]) : null;
  }

  /**
   * Find payment by Stripe payment intent ID
   */
  async findByStripePaymentIntentId(stripePaymentIntentId: string): Promise<Payment | null> {
    const { rows } = await this.db.query<any>(
      'SELECT * FROM payments WHERE stripe_payment_intent_id = $1',
      [stripePaymentIntentId],
    );

    return rows[0] ? mapToPayment(rows[0]) : null;
  }

  /**
   * List all payments for a user
   */
  async listByUser(
    userId: string,
    limit: number = 50,
    offset: number = 0,
  ): Promise<Payment[]> {
    const { rows } = await this.db.query<any>(
      `SELECT * FROM payments
       WHERE user_id = $1
       ORDER BY created_at DESC
       LIMIT $2 OFFSET $3`,
      [userId, limit, offset],
    );

    return rows.map(mapToPayment);
  }

  /**
   * Count total payments for a user
   */
  async countByUser(userId: string): Promise<number> {
    const { rows } = await this.db.query<{ count: string }>(
      'SELECT COUNT(*) as count FROM payments WHERE user_id = $1',
      [userId],
    );

    return parseInt(rows[0].count, 10);
  }

  /**
   * Create a new payment
   */
  async create(data: {
    userId: string;
    amount: number;
    currency: string;
    status: 'pending' | 'succeeded' | 'failed' | 'canceled';
    stripePaymentIntentId: string;
    reportId?: string;
  }): Promise<Payment> {
    const { rows } = await this.db.query<any>(
      `INSERT INTO payments (user_id, report_id, amount, currency, status, stripe_payment_intent_id)
       VALUES ($1, $2, $3, $4, $5, $6)
       RETURNING *`,
      [
        data.userId,
        data.reportId ?? null,
        data.amount,
        data.currency,
        data.status,
        data.stripePaymentIntentId,
      ],
    );

    return mapToPayment(rows[0]);
  }

  /**
   * Update payment status
   */
  async updateStatus(
    paymentId: string,
    status: 'pending' | 'succeeded' | 'failed' | 'canceled',
  ): Promise<void> {
    await this.db.query(
      'UPDATE payments SET status = $1 WHERE payment_id = $2',
      [status, paymentId],
    );
  }

  /**
   * Link payment to report
   */
  async linkToReport(paymentId: string, reportId: string): Promise<void> {
    await this.db.query(
      'UPDATE payments SET report_id = $1 WHERE payment_id = $2',
      [reportId, paymentId],
    );
  }

  /**
   * List successful payments for a user
   */
  async listSuccessfulByUser(userId: string): Promise<Payment[]> {
    const { rows } = await this.db.query<any>(
      `SELECT * FROM payments
       WHERE user_id = $1
         AND status = 'succeeded'
       ORDER BY created_at DESC`,
      [userId],
    );

    return rows.map(mapToPayment);
  }

  /**
   * Count successful payments for a user
   */
  async countSuccessfulByUser(userId: string): Promise<number> {
    const { rows } = await this.db.query<{ count: string }>(
      `SELECT COUNT(*) as count FROM payments
       WHERE user_id = $1
         AND status = 'succeeded'`,
      [userId],
    );

    return parseInt(rows[0].count, 10);
  }

  /**
   * Get total revenue (sum of all successful payments)
   */
  async getTotalRevenue(): Promise<number> {
    const { rows } = await this.db.query<{ total: string }>(
      `SELECT COALESCE(SUM(amount), 0) as total FROM payments
       WHERE status = 'succeeded'`,
    );

    return parseInt(rows[0].total, 10);
  }

  /**
   * Check if user has any successful payment
   */
  async hasSuccessfulPayment(userId: string): Promise<boolean> {
    const { rowCount } = await this.db.query(
      `SELECT 1 FROM payments
       WHERE user_id = $1
         AND status = 'succeeded'
       LIMIT 1`,
      [userId],
    );

    return rowCount > 0;
  }
}
